src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/SessionTemplateComparator.kt

```diff
@@ -0,0 +1,206 @@
+
+  private val validMatch = object : BiPredicate<String?, String?> {
+    override fun test(permittedSessionLevel: String?, prisonerLevel: String?): Boolean {
+      permittedSessionLevel?.let {
+        return it == prisonerLevel
+      }
+      // If no prison level then match
+      return true
+    }
+  }
+
+  private val multipleLocationMatcher =
+    BiPredicate<PermittedSessionLocationDto, Set<PermittedSessionLocationDto>> { primarySessionLocation, comparedSessionLocations ->
+      comparedSessionLocations.stream().anyMatch { comparedSessionLocation ->
+        validMatch.test(primarySessionLocation.levelOneCode, comparedSessionLocation.levelOneCode)
+          .and(validMatch.test(primarySessionLocation.levelTwoCode, comparedSessionLocation.levelTwoCode))
+          .and(validMatch.test(primarySessionLocation.levelThreeCode, comparedSessionLocation.levelThreeCode))
+          .and(validMatch.test(primarySessionLocation.levelFourCode, comparedSessionLocation.levelFourCode))
+      }
+    }
+
+  private fun hasOverlappingDates(
+    newSessionDetails: SessionDetailsDto,
+    existingSessionDetails: SessionDetailsDto,
+  ): Boolean {
+    val newFromDate = newSessionDetails.sessionDateRange.validFromDate
+    val existingFromDate = existingSessionDetails.sessionDateRange.validFromDate
+    val newToDate = newSessionDetails.sessionDateRange.validToDate ?: newFromDate.plusYears(500)
+    val existingToDate = existingSessionDetails.sessionDateRange.validToDate ?: newSessionDetails.sessionDateRange.validToDate?.plusDays(1) ?: newToDate
+
+    return (
+      // new from date between existing from and to dates
+      (newFromDate >= existingFromDate && newFromDate <= existingToDate) ||
+        // new to date between existing from and to dates
+        (newToDate >= existingFromDate && newToDate <= existingToDate) ||
+        // new from date less than existing from date and new to date greater than existing to date
+        (newFromDate <= existingFromDate && newToDate >= existingToDate)
+      )
+  }
+
+  private fun hasOverlappingTimes(
+    newSessionDetails: SessionDetailsDto,
+    existingSessionDetails: SessionDetailsDto,
+  ): Boolean {
+    val newStartTime = newSessionDetails.sessionTimeSlot.startTime
+    val existingStartTime = existingSessionDetails.sessionTimeSlot.startTime
+    val newEndTime = newSessionDetails.sessionTimeSlot.endTime
+    val existingEndTime = existingSessionDetails.sessionTimeSlot.endTime
+
+    return (
+      // new start time between existing start and end times
+      (newStartTime >= existingStartTime && newStartTime <= existingEndTime) ||
+        // new end time between existing start and end times
+        (newEndTime >= existingStartTime && newEndTime <= existingEndTime) ||
+        // new start time less than existing start time and new end time greater than existing end time
+        (newStartTime <= existingStartTime && newEndTime >= existingEndTime)
```

Comment:
Use in **in** operator

Reply:
```
    return (
      // new start time between existing start and end times
      (newStartTime in existingStartTime..existingEndTime) ||
        // new end time between existing start and end times
        (newEndTime in existingStartTime..existingEndTime) ||
        // new start time less than existing start time and new end time greater than existing end time
        (newStartTime <= existingStartTime && newEndTime >= existingEndTime)
      )
```

Reply:
can use in here but again will it be less performative?

Reply:
don't think so.... no reason why it would be